vllm/entrypoints/openai/tool_parsers/internlm2_tool_parser.py

```diff
@@ -0,0 +1,122 @@
+import json
+from typing import Sequence, Union
+
+from vllm.entrypoints.openai.protocol import (ChatCompletionRequest,
+                                              DeltaFunctionCall, DeltaMessage,
+                                              DeltaToolCall,
+                                              ExtractedToolCallInformation,
+                                              FunctionCall, ToolCall)
+from vllm.entrypoints.openai.tool_parsers.abstract_tool_parser import (
+    ToolParser, ToolParserManager)
+from vllm.logger import init_logger
+from vllm.transformers_utils.tokenizer import AnyTokenizer
+from vllm.utils import random_uuid
+
+logger = init_logger(__name__)
+
+
+@ToolParserManager.register_module(["internlm2", "internlm2_5"])
+class Internlm2ToolParser(ToolParser):
+
+    def __init__(self, tokenizer: AnyTokenizer):
+        super().__init__(tokenizer)
+        self.position = 0
+        self.current_tool_id = 0
+
+    def adjust_request(
+            self, request: ChatCompletionRequest) -> ChatCompletionRequest:
+        if request.tools and request.tool_choice != 'none':
+            # do not skip special tokens because internlm use the special
+            # tokens to indicated the start and end of the tool calls
+            # information.
+            request.skip_special_tokens = False
+        return request
+
+    def extract_tool_calls_streaming(
+        self,
+        previous_text: str,
+        current_text: str,
+        delta_text: str,
+        previous_token_ids: Sequence[int],
+        current_token_ids: Sequence[int],
+        delta_token_ids: Sequence[int],
+        request: ChatCompletionRequest,
+    ) -> Union[DeltaMessage, None]:
```

Comment:
This streaming implementation is very different from what a typical one would be - it's kind of "naive" in that it waits until the _entire_ tool call has been generated and then streams the name and the entire arguments at once, instead of streaming the function name once it's generated, and then argument diffs as the arguments are generated. 

This is off-spec for OpenAI tool streaming.

It may very well still work for some use-cases, but you will receive information about the called tool's name much later than you otherwise would, and your stream is going to look very unlike what it would from OpenAI or from another tool call parser.

Reply:
I can take a look at this and see if I can help make it better this weekend

Reply:
I'm working on this now,  after I finish this, may be you can take a check on this.

Reply:
it's finished, you can take a look of this.

Reply:
ok, I will take a look this weekend!